src/Microsoft.Identity.Client/UI/CustomWebUiHandler.cs

```diff
@@ -0,0 +1,107 @@
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.UI
+{
+    internal class CustomWebUiHandler : IWebUI
+    {
+        private readonly ICustomWebUi _customWebUi;
+
+        public CustomWebUiHandler(ICustomWebUi customWebUi)
+        {
+            _customWebUi = customWebUi;
+        }
+
+        /// <inheritdoc />
+        public async Task<AuthorizationResult> AcquireAuthorizationAsync(
+            Uri authorizationUri,
+            Uri redirectUri,
+            RequestContext requestContext)
+        {
+            requestContext.Logger.Info(LogMessages.CustomWebUiAcquiringAuthorizationCode);
+
+            try
+            {
+                var uri = await _customWebUi.AcquireAuthorizationCodeAsync(authorizationUri, redirectUri)
+                                            .ConfigureAwait(false);
+                if (uri == null)
+                {
+                    throw new MsalCustomWebUiFailedException(CoreErrorMessages.CustomWebUiReturnedNullUri);
+                }
+
+                if (uri.Authority.Equals(redirectUri.Authority, StringComparison.OrdinalIgnoreCase) &&
+                    uri.AbsolutePath.Equals(redirectUri.AbsolutePath))
+                {
+                    IDictionary<string, string> inputQp = CoreHelpers.ParseKeyValueList(
+                        authorizationUri.Query.Substring(1),
+                        '&',
+                        true,
+                        null);
+
+                    requestContext.Logger.Info(LogMessages.CustomWebUiRedirectUriMatched);
+                    return new AuthorizationResult(AuthorizationStatus.Success, uri.OriginalString)
+                    {
+                        State = inputQp[OAuth2Parameter.State]
+                    };
+                }
+
+                throw new MsalCustomWebUiFailedException(CoreErrorMessages.CustomWebUiRedirectUriWasNotMatchedToProperUri);
```

Comment:
>CoreErrorMessages.CustomWebUiRedirectUriWasNotMatchedToProperUri [](start = 57, length = 64)

consider adding the redirect uri specified which didn't match?

Reply:
done